File: src/main/java/com/vaadin/demo/component/textfield/TextFieldBasicFeatures.java
package com.vaadin.demo.component.textfield;

import com.vaadin.flow.component.html.Span;
import com.vaadin.flow.component.icon.VaadinIcon;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.component.textfield.TextField;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("text-field-basic-features")
public class TextFieldBasicFeatures extends HorizontalLayout {

    public TextFieldBasicFeatures() {
        setPadding(false);

        // tag::snippet[]
        TextField field = new TextField();
        field.setLabel("Label");
        field.setHelperText("Helper text");
        field.setPlaceholder("Placeholder");
        field.setTooltipText("Tooltip text");
        field.setClearButtonVisible(true);
        field.setPrefixComponent(VaadinIcon.VAADIN_H.create());
        field.setSuffixComponent(new Span(":)"));
        // end::snippet[]

        add(field);
    }

    public static class Exporter extends DemoExporter<TextFieldBasicFeatures> { // hidden-source-line
    } // hidden-source-line
}

File: frontend/demo/component/textfield/text-field-basic-features.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/text-field';
import '@vaadin/icon';
import '@vaadin/icons';
import '@vaadin/tooltip';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('text-field-basic-features')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-text-field
        label="Label"
        helper-text="Helper text"
        placeholder="Placeholder"
        clear-button-visible
      >
        <vaadin-tooltip slot="tooltip" text="Tooltip text"></vaadin-tooltip>
        <vaadin-icon slot="prefix" icon="vaadin:vaadin-h"></vaadin-icon>
        <span slot="suffix">:)</span>
      </vaadin-text-field>
      <!-- end::snippet[] -->
    `;
  }
}
